odoo/service/server.py

```diff
@@ -1041,16 +1092,119 @@ def start(self):
+            self.multi.socket.close()
+
+    def _db_list(self):
+        if config['db_name']:
+            db_names = config['db_name'].split(',')
+        else:
+            db_names = odoo.service.db.list_dbs(True)
+        return db_names
+
+    def process_work(self):
+        rpc_request = logging.getLogger('odoo.netsvc.rpc.request')
+        rpc_request_flag = rpc_request.isEnabledFor(logging.DEBUG)
+        _logger.debug("%s (%s) polling for jobs", self.__class__.__name__, self.pid)
+
+        db_names = self._db_list()
+        if not db_names:
+            self.db_index = 0
+            return
+
+        db_name = self._next_db(db_names)
+        if not db_name:
+            self.db_index = 0
+            return
+
+        self.setproctitle(db_name)
+        if rpc_request_flag:
+            start_time = time.time()
+            start_memory = memory_info(psutil.Process(os.getpid()))
+
+        self._process_work(db_name)
+        # dont keep cursors in multi database mode
+        if len(db_names) > 1:
+            odoo.sql_db.close_db(db_name)
+
+        if rpc_request_flag:
+            run_time = time.time() - start_time
+            end_memory = memory_info(psutil.Process(os.getpid()))
+            vms_diff = (end_memory - start_memory) / 1024
+            logline = '%s time:%.3fs mem: %sk -> %sk (diff: %sk)' % \
+                (db_name, run_time, start_memory / 1024, end_memory / 1024, vms_diff)
+            _logger.debug("%s (%s) %s", self.__class__.__name__, self.pid, logline)
+
+        self.request_count += 1
+        if self.request_count >= self.request_max and self.request_max < len(db_names):
+            _logger.error("There are more dabatases to process than allowed "
+                          "by the `limit_request` configuration variable: %s more.",
+                          len(db_names) - self.request_max)
+
+    def _process_work(self, db_name):
+        pass
+
+    def _next_db(self):
+        pass
+
+
+class WorkerAsync(WorkerJobCommon):
+    def __init__(self, multi):
+        from odoo.addons.base.models.ir_async import ir_async
+        super(WorkerAsync, self).__init__(multi)
+        self.watchdog_timeout = multi.async_timeout
+        self._process_work = ir_async._process_jobs
+
+    def start(self):
+        super().start()
+        self.conn = psycopg2.connect(**connection_info_for("postgres")[1])
+        with self.conn.cursor() as cr:
+            cr.execute("LISTEN odoo_async;")
+        self.conn.commit()
+
+        self.notified_dbs = [
+            dbname
+            for dbname, registry
+            in odoo.modules.registry.Registry.registries.items()
+            if registry.ready
+        ]
+
+    def stop(self):
+        self.conn.close()
+        super().close()
+
+    def sleep(self):
+        if not self.notified_dbs:
+            select.select([self.conn, self.wakeup_fd_r], [], [], SLEEP_INTERVAL)
+            time.sleep(self.pid / 100 % .1)  # avoid thunderhead effect
```

Comment:
Why does this sleep after sleeping?

Reply:
All workers are awaken at the same time, the sleep prevents they all poll the database at the exact same time. Shall I remove it ?

Reply:
Ah I see, makes sense.

Reply:
Simply add what you wrote in a comment in the code.
That will definitively help reading it!